### ISSUE TITLE
Patched Terminess fonts still blurry

### ISSUE THREAD

**Issue by Slabity** [ryanoasis/nerd-fonts#1329]
#### 🗹 Requirements

- [X] I have searched the [issues](https://github.com/ryanoasis/nerd-fonts/issues) for my issue and found nothing related and/or helpful
- [X] I have searched the [FAQ](https://github.com/ryanoasis/nerd-fonts/wiki/FAQ-and-Troubleshooting) for help
- [X] I have searched the [Wiki](https://github.com/ryanoasis/nerd-fonts/wiki) for help

(Note that there are some older issues reported for this, but are locked for being stale)

#### 🎯 Subject of the issue

Experienced behavior:
Decided to try using the patched Terminess fonts again since #329 was patched, but the font still looks blurry.

Here's a screenshot of the current, patched font at 10px:

![image](https://github.com/ryanoasis/nerd-fonts/assets/3788421/02e20312-be8a-4b60-a833-522c5d39ce49)

Expected behavior:

Here's a screenshot of the unpatched font, also at 10px:

![image](https://github.com/ryanoasis/nerd-fonts/assets/3788421/538f7f24-9b8d-4d55-8df1-ff586a6d29c0)

(Note the size doesn't affect the blurriness, it's blurry at all sizes).

Also, it looks fine when I patch it manually myself (this one is 8px though):

![image](https://github.com/ryanoasis/nerd-fonts/assets/3788421/cfd9b57d-2936-4fe5-a32b-9377020e3ee6)

Example symbols:
Pretty much all of them.

#### 🔧 Your Setup

- _Which font are you using (e.g. `Anonymice Powerline Nerd Font Complete.ttf`)?_
   - `TerminessNerdFont-Regular.ttf`
- _Where did you get the file from (download link, self patched, source downloaded from link...)_
   - Cloned directly from here
- _Which terminal emulator are you using (e.g. `iterm2`, `urxvt`, `gnome`, `konsole`)?_
   - `alacritty`, but screenshots are from font-manager
- _Are you using OS X, Linux or Windows? And which specific version or distribution?_
   - NixOS, a Linux distribution

#### ★ Screenshots (Optional)

Posted above in behavior sections

**Comment by Slabity** [comment]
I'm mostly creating this issue because I can patch the fonts manually, but I can't seem to figure out how to fix the actual repo to get clean bitmap fonts. I think it might be the same issue as #1174 for AnonymousPro, but I'm not sure what the fix in #1175 actually does to resolve it.

I wouldn't be surprised if a few other bitmap fonts are having the same issue.

**Comment by Finii** [comment]
Thanks for reporting. Can you provide the output of the patcher with `--debug 2` which will help me see some detail.

> I'm mostly creating this issue because I can patch the fonts manually

How do you do that?

The AnonymousPro issue was that we used a font without bitmaps as source, of course there are no bitmaps than in the patched version. Changing to a AnonymousPro-with-bitmaps resulted in a patched AnonymousPro with bitmaps :grimacing: 

The unpatched Terminus source we use seems to have bitmaps...:

![image](https://github.com/ryanoasis/nerd-fonts/assets/16012374/dcd32f3f-7181-42df-ad28-8ba46050cf1e)

and so does the patched font...:

![image](https://github.com/ryanoasis/nerd-fonts/assets/16012374/0b698dd0-f393-41b2-a62c-644bcf0bd88b)

**Comment by Slabity** [comment]
@Finii - Sure, I can replicate this by using the `ter-u12n.otb` file that came in the `terminus-font-4.49.1` package that NixOS provides and just ran `nerd-font-patcher ter-u12n.otb` without any extra options:

```shell
[nix-shell:~]$ cp /nix/store/r3hizlibf6bv79ydaxyd9chfsp1783yl-terminus-font-4.49.1/share/fonts/terminus/ter-u12n.otb .

[nix-shell:~]$ chmod +w ter-u12n.otb

[nix-shell:~]$ nerd-font-patcher ter-u12n.otb
No configfile given, skipping configfile related actions
Nerd Fonts Patcher v2.2.2 (3.0.5) executing

Adding 59 Glyphs from Seti-UI + Custom Set
╢████████████████████████████████████████╟ 100%
Adding 198 Glyphs from Devicons Set
╢████████████████████████████████████████╟ 100%

Done with Patch Sets, generating font...

Changing lowestRecPPEM from 8 to 12

Generated: Terminess Regular Nerd Font in './Terminess Regular Nerd Font.otb'
```

And the result is this:

![image](https://github.com/ryanoasis/nerd-fonts/assets/3788421/3be55e3b-df16-4a5f-b478-e51bd2a9a0ce)

Note that I have tried playing with my system's fontconfig settings and toggling `anti-aliasing` and `hinting` in any combination didn't make it any better.

**Comment by Slabity** [comment]
Alright, so apparently we did have 3.0.2 in Nixpkgs, but I mistakenly used an out of date version. Here is the output of the current one:

```bash
$ nerd-font-patcher --debug 2 ter-u12n.otb
Nerd Fonts Patcher v3.0.2 (4.4.0) (ff 20230101)
DEBUG: Naming mode 1
DEBUG: Monospace check: Panose says "monospaced"; glyph-width-mono True
DEBUG: Final font cell dimensions 512 w x 1024 h
DEBUG: 150/160 box drawing glyphs will be replaced
Adding 180 Glyphs from Seti-UI + Custom Set
╢████████████████████████████████████████╟ 100%
Adding 6 Glyphs from Heavy Angle Brackets Set
╢████████████████████████████████████████╟ 100%
Adding 160 Glyphs from Box Drawing Set
╢████████████████████████████████████████╟ 100%
Adding 198 Glyphs from Devicons Set
╢████████████████████████████████████████╟ 100%
Done with Patch Sets, generating font...
DEBUG: =====> Family (ID 1)      ok       (19 <=31): Terminess Nerd Font
DEBUG: =====> SubFamily (ID 2)   ok       ( 7 <=31): Regular
DEBUG: =====> Fullname (ID 4)    ok       (19 <=63): Terminess Nerd Font
DEBUG: =====> PSN (ID 6)         ok       (19 <=63): TerminessNF-Regular
DEBUG: Preserving bitmaps (12,)
   Terminess Nerd Font
   \===> 'TerminessNerdFont-Regular.otb'
```

Oddly enough, *now* the output seems to be blurry:

![image](https://github.com/ryanoasis/nerd-fonts/assets/3788421/190d1af1-4a97-41f7-8ae1-cf1968448db7)

I do notice that the font sizes are not integer based. For some reason I can only select 8.2, 8.7, 9.2, 9.7, 10.2, etc. Perhaps related?

**Comment by Finii** [comment]
Thanks for the investigation!

> I do notice that the font sizes are not integer based. For some reason I can only select 8.2, 8.7, 9.2, 9.7, 10.2, etc

That is, in `font manager`? Which font-manager is it, I believe there are multiple different almost same-named ones ;-)

I must confess I personally never use bitmap fonts and my knowledge about bitmap fonts is centuries old (i.e. before ttf and / or when X was more or less bitmap fonts only). But I will try to find something out. In the light of the changed selectable sizes I fear it might be `fontforge` has yet another bug.

**Comment by Finii** [comment]
Hmm, looks similar but slightly different :grimacing: 

![image](https://github.com/ryanoasis/nerd-fonts/assets/16012374/4a43a4d3-d5e3-407c-a17c-1444a1f0167b)

**Comment by Slabity** [comment]
The `font-manager` [package in Nixpkgs](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/misc/font-manager/default.nix) appears to be version 0.8.8 of [https://github.com/FontManager/font-manager/](FontManager/font-manager). I think all of those screenshots might be the same app with different themes. I'm specifically using `arc-dark`, while the two you have might be the default light and dark themes.

Though I can confirm the same result in `alacritty` as well, so it's not just the GTK applications. Here's the two versions side-by-side:

![image](https://github.com/ryanoasis/nerd-fonts/assets/3788421/2007a4d6-3384-4c94-9771-cab9d7bdda8c)

There's a difference in both the characters and the spacing from the looks of it. I'm not certain both `font-manager` and `alacritty` use the same font-rendering backend, but if you know of another app I should try I can do that as well.

**Comment by Slabity** [comment]
Just to add another datapoint for a different font renderer, here is Firefox rendering the different versions of the font. Only the Nerd Fonts version has the blur:

![image](https://github.com/ryanoasis/nerd-fonts/assets/3788421/e69b88bd-5416-428b-b996-2cf3faca44db)

Not sure what other apps I can use to see if it's a rendering issue or not. Unfortunately my knowledge of how fonts, font rendering, and how the font patcher works is next to nothing.

**Comment by Finii** [comment]
Thanks for the feedback @herself I will look into it before 3.2.

**Comment by Finii** [comment]
@allcontributors please add @Slabity for bug

**Comment by Finii** [comment]
In fact I do not really know if that solves @Slabity blurryness problem.
At least the outlines are broken in the version we had.

It can also be a hinting problem, if the embedded bitmaps are not used.
Ah, yes, the font seems to be unhinted:

![image](https://github.com/ryanoasis/nerd-fonts/assets/16012374/79cb55c0-e916-4a40-b289-9602d8f7338f)

If your problem persists I can try and throw `ttfautohint` unto the font, and then you can try that font.
Please let me know :-)

**Comment by Finii** [comment]
Ah yes...

![image](https://github.com/ryanoasis/nerd-fonts/assets/16012374/3d5fc7a1-7c85-4cee-a6d7-4439afb1330b)

Hmm, but ... of course the original also has no hints, so that should not be the problem.
Maybe the bitmap part of the font is still broken.
And/or the renderer switches to outlines if not all glyphs are available as bitmaps.
My knowledge of bitmap fonts is not very good (next to nonexisting)

You use `alacritty`, I guess I have that installed (i.e. have it self-built somewhere, lets see)

![image](https://github.com/ryanoasis/nerd-fonts/assets/16012374/235fa558-f1bf-4777-b9e9-b72002bf3604)

## testing

On my machine even the unpatched font is unsharp in `alacritty`:

![image](https://github.com/ryanoasis/nerd-fonts/assets/16012374/f9b3ab9e-0e66-45c4-979c-8b856a807cce)

Interestingly `alacritty` ignores the system wide hinting and antialiasing setting. Turned off everything system wide, see how bad the window titles and `tilix` text window looks now, but `alacritty` does what it wants :roll_eyes: 

![image](https://github.com/ryanoasis/nerd-fonts/assets/16012374/3766bb59-dc25-45f2-b562-6bc5a0edf9ae)

**tbc**

**Comment by Finii** [comment]
Alacritty seems to have 'a problem' when a font is not all-bitmaps it will switch to LCD subpixel rendering and there is no way to prevent that?

See for example
* https://github.com/alacritty/alacritty/issues/3459

Maybe they fixed it in the meantime, but not having an option to influence antialiasing (subpixel rendering) is a strange choice.

If this is the problem we can not do anything about it. You can patch your `alacritty` source like described in the linked Issue, to turn the unwanted antialiasing off and get sharp glyphs back. If that works at all.

I guess bitmap fonts and patching just does not blend very well.

**Comment by Slabity** [comment]
@Finii - I don't think this is fixed. I'm seeing the same results on v3.2.1

**Comment by rck** [comment]
@Finii I also think this isn't fixed in v3.2.1. Noticed it on alacritty, but I see the same in Gnome "console".
